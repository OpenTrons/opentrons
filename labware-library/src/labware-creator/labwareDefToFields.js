// @flow
import flatten from 'lodash/flatten'
import isEqual from 'lodash/isEqual'
import round from 'lodash/round'
import omit from 'lodash/omit'
import { getSpacing } from '../labwareInference'
import type { LabwareDefinition2 } from '@opentrons/shared-data'
import type { LabwareFields, BooleanString } from './fields'

const boolToBoolString = (b: boolean): BooleanString => (b ? 'true' : 'false') // TODO IMMEDIATELY revisit, is this duplicated elsewhere?

// NOTE: this fn should always be passed an object valid under JSON schema for LabwareDefinition2
// TODO: Ian 2019-08-26 this shares functionality with getUniqueWellProperties fn in labware-library/src/definitions.js
export default function labwareDefToFields(
  def: LabwareDefinition2
): ?LabwareFields {
  const groups = def.groups || []
  const allWells = flatten(def.ordering).map(wellName => def.wells[wellName])
  // single well we'll use to represent the entire labware if homogeneousWell is true
  const commonWell = def.wells.A1

  const wellDimKeys = ['x', 'y', 'z']
  const commonWellNoDims = omit(commonWell, wellDimKeys)
  const homogeneousWells = allWells.every(well =>
    isEqual(omit(well, wellDimKeys), commonWellNoDims)
  )

  // don't bother trying to infer row/col regularity, assume `groups` is correct. If `groups` is missing, assume irregular
  const gridSpacingX = getSpacing(allWells, 'x')
  const gridSpacingY = getSpacing(allWells, 'y')
  const regularColumnSpacing = gridSpacingX !== null
  const regularRowSpacing = gridSpacingY !== null

  if (!homogeneousWells || !regularRowSpacing || !regularColumnSpacing) {
    // TODO IMMEDIATELY: somehow handle uploading "irregular" (multi-grid) labware, error messaging
    console.warn('TODO! unhandled labware def', {
      homogeneousWells,
      regularColumnSpacing,
      regularRowSpacing,
    })
    return null
  }

  // use first group, if it exists
  const group = groups[0] || null

  const gridRowsNum = def.ordering[0].length
  const gridColumnsNum = def.ordering.length

  return {
    // NOTE: Ian 2019-08-26 these fields cannot be inferred
    labwareType: null,
    tubeRackInsertLoadName: null,
    aluminumBlockType: null,
    aluminumBlockChildType: null,

    footprintXDimension: String(def.dimensions.xDimension),
    footprintYDimension: String(def.dimensions.yDimension),
    labwareZDimension: String(def.dimensions.zDimension),

    // Missing values or zeroes for these fields should be translated to null
    gridRows: gridRowsNum > 0 ? String(gridRowsNum) : null,
    gridColumns: gridColumnsNum > 0 ? String(gridColumnsNum) : null,
    gridSpacingX:
      gridSpacingX == null || gridSpacingX === 0 ? null : String(gridSpacingX),
    gridSpacingY:
      gridSpacingY == null || gridSpacingY === 0 ? null : String(gridSpacingY),

    gridOffsetX: String(commonWell.x),
    // y offset is dist btw labware 'top' (furthest in y) and first well y
    gridOffsetY: String(round(def.dimensions.yDimension - commonWell.y, 2)),

    homogeneousWells: boolToBoolString(homogeneousWells),
    regularRowSpacing: boolToBoolString(regularRowSpacing),
    regularColumnSpacing: boolToBoolString(regularColumnSpacing),

    wellVolume: String(commonWell.totalLiquidVolume),
    wellBottomShape: group?.metadata?.wellBottomShape || null,
    wellDepth: String(commonWell.depth),
    wellShape: commonWell.shape,

    // used with circular well shape only
    wellDiameter:
      commonWell.shape === 'circular' ? String(commonWell.diameter) : null,

    // used with rectangular well shape only
    wellXDimension:
      commonWell.shape === 'rectangular' ? String(commonWell.xDimension) : null,
    wellYDimension:
      commonWell.shape === 'rectangular' ? String(commonWell.yDimension) : null,

    brand: def.brand.brand,
    brandId: def.brand.brandId ? def.brand.brandId.join(',') : null, // comma-separated values

    loadName: def.parameters.loadName,
    displayName: def.metadata.displayName,

    // fields for test protocol
    pipetteName: null,
  }
}
