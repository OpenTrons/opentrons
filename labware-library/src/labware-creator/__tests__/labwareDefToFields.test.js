// @flow
import labwareDefToFields from '../labwareDefToFields'
import fixtureRegularExample1 from '@opentrons/shared-data/labware/fixtures/2/fixture_regular_example_1'
import fixtureIrregularExample1 from '@opentrons/shared-data/labware/fixtures/2/fixture_irregular_example_1'

jest.mock('../../definitions')

describe('labwareDefToFields', () => {
  test('fixture_regular_example_1', () => {
    const def = fixtureRegularExample1
    const result = labwareDefToFields(def)
    expect(result).toEqual({
      labwareType: null,
      tubeRackInsertLoadName: null,
      aluminumBlockType: null,
      aluminumBlockChildType: null,

      footprintXDimension: String(def.dimensions.xDimension),
      footprintYDimension: String(def.dimensions.yDimension),
      labwareZDimension: String(def.dimensions.zDimension),

      gridRows: '1',
      gridColumns: '2',
      gridSpacingX: '10',
      gridSpacingY: null, // only single row, no Y spacing
      gridOffsetX: '10',
      gridOffsetY: '75.48',

      homogeneousWells: 'true',
      regularRowSpacing: 'true',
      regularColumnSpacing: 'true',

      wellVolume: '100',
      wellBottomShape: null, // TODO IMMEDIATELY: rethink this field?
      wellDepth: '40',
      wellShape: 'circular',

      wellDiameter: '30',

      // used with rectangular well shape only
      wellXDimension: null,
      wellYDimension: null,

      brand: 'opentrons',
      brandId: 't40u9sernisofsea',

      loadName: def.parameters.loadName,
      displayName: def.metadata.displayName,

      pipetteName: null,
    })
  })

  test('fixture_irregular_example_1 should return null (not yet supported)', () => {
    const def = fixtureIrregularExample1
    const result = labwareDefToFields(def)
    expect(result).toEqual(null)
  })
})
