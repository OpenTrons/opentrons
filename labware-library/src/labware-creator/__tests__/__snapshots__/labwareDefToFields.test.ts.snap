// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`labwareDefToFields fixture_12_trough 1`] = `
Object {
  "aluminumBlockChildType": null,
  "aluminumBlockType": null,
  "brand": "USA Scientific",
  "brandId": "1061-8150",
  "displayName": null,
  "footprintXDimension": "127.76",
  "footprintYDimension": "85.8",
  "gridColumns": "12",
  "gridOffsetX": "13.94",
  "gridOffsetY": "42.9",
  "gridRows": "1",
  "gridSpacingX": "9.09",
  "gridSpacingY": null,
  "handPlacedTipFit": null,
  "homogeneousWells": "true",
  "labwareType": "reservoir",
  "labwareZDimension": "44.45",
  "loadName": null,
  "pipetteName": null,
  "regularColumnSpacing": "true",
  "regularRowSpacing": "true",
  "tubeRackInsertLoadName": null,
  "wellBottomShape": "v",
  "wellDepth": "42.16",
  "wellDiameter": null,
  "wellShape": "rectangular",
  "wellVolume": "22000",
  "wellXDimension": "8.33",
  "wellYDimension": "71.88",
}
`;
